### PR TITLE
Improve `ExpressionEvaluator`

### PR DESCRIPTION
**Pull request by zasdfgbnm**
This PR improves `ExpressionEvaluator` by doing the following two things:
1. Add a boolean flag `evaluate_validate` to `ExpressionEvaluator::bind`. This flag is only related when binding an intermediate value. When set to true, then if all the dependencies of the given `Val*` is already satisfied, the expr evaluator will actually run an evaluation and check if the evaluated value match with the given value. When set to false, the expr evaluator will assume the given value is correct and just bind it.
2. When binding a `TensorView`, also bind its sizes.

With this two change, I can just remove the `bindInputForExprEvaluation`. The validation in `bind` should be stricter than the original check in `bindInputForExprEvaluation`, because I do find a bug in `FusionMagicSchedulerInstanceNormalizationBackward_CUDA`, where there are two inputs with the same `Val*` but are provided different runtime values. For such kind of errors, our system can now detect it easily.

### CODE REVIEW

File: test/test_gpu_tensorcore.cpp [hunk 1]
```diff
@@ -4386,7 +4386,7 @@ TEST_F(NVFuserTest, FusionMatmulSchedulerEpilogueAlphaBetaBias_CUDA) {
 
   // NOTE: increasted absolute tolerance to silence false negative verification
   //       caused by different way of calculating reference
-  TORCH_CHECK(outputs[0].allclose(t8, 0.001, 0.001));
+  TORCH_CHECK(outputs[0].allclose(t8, 0.01, 0.01));
```
Comment (zasdfgbnm) [inline test/test_gpu_tensorcore.cpp:4389]: Unrelated test fix

**Comment by zasdfgbnm** [comment]
!build

File: csrc/expr_evaluator.cpp [hunk 2]
```diff
@@ -79,15 +80,62 @@ void validateValWithConcreteValue(
...
+    auto same = is_same(evaluated_value, concrete_value);
+    TORCH_CHECK(
+        same,
+        "Tried to bind to a value: ",
+        value->toInlineString(),
+        "(which evaluated to ",
+        evaluated_value,
+        ") as ",
+        concrete_value);
+  }
```
[the remark this answers is not in the document]
Comment (zasdfgbnm) [reply csrc/expr_evaluator.cpp:119]: Good catch! Added a `toString` in `PolymorphicValue_functions` and uses that function for error messages.

File: csrc/polymorphic_value.h [hunk 3]
```diff
@@ -255,6 +255,14 @@ using PolymorphicValue = DynamicType<
 
 namespace PolymorphicValue_functions {
 
+inline bool is_same(const PolymorphicValue& a, const PolymorphicValue& b) {
+  if (a.is<at::Tensor>() && b.is<at::Tensor>()) {
+    return (a.as<at::Tensor>().is_same(b.as<at::Tensor>()));
+  } else {
+    return (a == b);
```
[the remark this answers is not in the document]
Comment (zasdfgbnm) [reply csrc/polymorphic_value.h:278]: Moved `validateValWithConcreteValue` above for better error message. Also, added a type check and return false if type mismatch.

File: csrc/expr_evaluator.cpp [hunk 4]
```diff
@@ -79,15 +80,62 @@ void validateValWithConcreteValue(
 
 } // namespace
 
+bool ExpressionEvaluator::readyToEvaluate(const Val* value) const {
```
[the remark this answers is not in the document]
Comment (zasdfgbnm) [reply csrc/expr_evaluator.cpp:83]: Changed `dependenciesSatisfied` into a template class so we can use `known_values_` here directly

[the remark this answers is not in the document]
Comment (zasdfgbnm) [reply csrc/expr_evaluator.cpp]: And also removed `readyToEvaluate` because it right now is very short

**Comment by zasdfgbnm** [comment]
!build

**Comment by zasdfgbnm** [comment]
!build

File: csrc/ir/utils.h [hunk 5]
```diff
@@ -427,16 +427,34 @@ void validateDomainEquivalence(
...
-      std::vector<const Val*>(needed_vals.begin(), needed_vals.end()),
-      std::unordered_set<const Val*>(known_vals.begin(), known_vals.end()));
+    ValOrVectorOfVal needed_vals,
+    const SetOfVal& known_vals = {}) {
+  if constexpr (
+      std::is_same_v<ValOrVectorOfVal, const Val*> ||
+      std::is_same_v<ValOrVectorOfVal, Val*>) {
+    return dependenciesSatisfied(
+        std::vector<const Val*>{needed_vals}, known_vals);
+  } else {
```
Comment (liqiangxl) [inline csrc/ir/utils.h:441]: can you add some comments? e.g. convert single val* to vector, default container is std::unordered_set<const Val*>>, what happens if other container is used, e.g. std::unordered_map.

**Review by liqiangxl — APPROVED** [review]
LGTM.

File: test/test_gpu_tensorcore.cpp (hunk 1 above)
Comment (zasdfgbnm) [reply test/test_gpu_tensorcore.cpp:4389]: added

File: csrc/ir/utils.h (hunk 5 above)
Comment (zasdfgbnm) [reply csrc/ir/utils.h:446]: added

**Comment by zasdfgbnm** [comment]
!build

**Comment by zasdfgbnm** [comment]
!build

**Review by naoyam — APPROVED** [review]
LGTM

File: test/test_gpu_tensorcore.cpp (hunk 1 above)
Comment (drzejan2) [reply test/test_gpu_tensorcore.cpp:4388]: On what device have you seen this test failing due to too high relative error?

Comment (drzejan2) [reply test/test_gpu_tensorcore.cpp:4388]: Ok, good to know - I will make sure to run tests against ad102 for new features I add from now on to avoid issues like this.

Thanks!
